src/logger.ts

```diff
@@ -11,75 +16,87 @@ const LEVEL = Object.freeze({
 });
 
 export default class Logger {
-  constructor (options = {}) {
-    this.formatter = options.formatter || Logger.colorFormatter;
-    this.destination = options.destination || process.stderr;
+  formatter: LogFormatter;
+  destination: NodeJS.WritableStream;
+  private _level: number;
+  history: LogEvent[];
+  private readonly _historySize: number;
+  constructor (options: {
+    formatter?: LogFormatter,
+    destination?: WriteStream,
+    historySize?: number,
+    level?: string
+  } = {}) {
+    this.formatter = options.formatter ?? Logger.colorFormatter;
+    this.destination = options.destination ?? process.stderr;
     this.history = [];
 
     this._historySize = options.historySize ?? 0;
-    this._level = LEVEL[process.env.MOJO_LOG_LEVEL || options.level || 'debug'];
+    this._level = LEVEL[process.env.MOJO_LOG_LEVEL ?? options.level ?? 'debug'];
   }
 
-  child (context) {
+  child (context: LogContext): ChildLogger {
     return new ChildLogger(this, context);
   }
 
-  static colorFormatter (data) {
+  static colorFormatter (data: LogEvent): string {
     const formatted = Logger.stringFormatter(data);
     if (data.level === 'error' || data.level === 'fatal') return chalk.red(formatted);
     if (data.level === 'warn') return chalk.yellow(formatted);
     return formatted;
   }
 
-  debug (msg, context) {
+  debug (msg: string, context?: LogContext): void {
     if (this._level >= LEVEL.debug) this._log('debug', msg, context);
   }
 
-  static stringFormatter (data) {
+  static stringFormatter (data: LogEvent): string {
     if (data.requestId !== undefined) return `[${data.time}] [${data.level}] [${data.requestId}] ${data.msg}\n`;
     return `[${data.time}] [${data.level}] ${data.msg}\n`;
   }
 
-  static systemdFormatter (data) {
+  static systemdFormatter (data: LogEvent): string {
     if (data.requestId === undefined) return `<${LEVEL[data.level]}>[${data.level.substring(0, 1)}] ${data.msg}\n`;
     return `<${LEVEL[data.level]}>[${data.level.substring(0, 1)}] [${data.requestId}] ${data.msg}\n`;
   }
 
-  error (msg, context) {
+  error (msg: string, context?: LogContext): void {
     if (this._level >= LEVEL.error) this._log('error', msg, context);
   }
 
-  fatal (msg, context) {
+  fatal (msg: string, context?: LogContext): void {
     if (this._level >= LEVEL.fatal) this._log('fatal', msg, context);
   }
 
-  info (msg, context) {
+  info (msg: string, context?: LogContext): void {
     if (this._level >= LEVEL.info) this._log('info', msg, context);
   }
 
-  static jsonFormatter (data) {
+  static jsonFormatter (data: LogEvent): string {
     return JSON.stringify(data);
   }
 
-  get level () {
-    return Object.keys(LEVEL).find(key => LEVEL[key] === this._level);
+  get level (): string {
+    const level = Object.keys(LEVEL).find(key => LEVEL[key] === this._level);
+    assert(level !== undefined, `Unsupported log level "${this._level}"`);
```

Comment:
This assertion needs a test case.

Reply:
changed to `as string` since setter already has the assert.